kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/BehaviorSpecGivenContainerScope.kt

```diff
@@ -94,4 +106,71 @@ class BehaviorSpecGivenContainerScope(
    private suspend fun addThen(name: String, test: suspend TestScope.() -> Unit, xdisabled: Boolean) {
       registerTest(TestName("Then: ", name, true), disabled = xdisabled, null, test)
    }
+
+   // data-test DSL follows
+
+   /**
+    * Registers tests inside the given test context for each element.
+    * The test name will be generated from the stable properties of the elements. See [StableIdents].
+    */
+   suspend fun <T> withData(
+      first: T,
+      second: T, // we need second to help the compiler disambiguate between this and the sequence version
+      vararg rest: T,
+      test: suspend BehaviorSpecGivenContainerScope.(T) -> Unit
+   ) = withData(listOf(first, second) + rest, test)
+
+   /**
+    * Registers tests inside the given test context for each element of [ts].
+    * The test names will be generated from the stable properties of the elements. See [StableIdents].
+    */
+   suspend fun <T> withData(
+      ts: Sequence<T>,
+      test: suspend BehaviorSpecGivenContainerScope.(T) -> Unit
+   ) = withData(ts.toList(), test)
+
+   /**
+    * Registers tests inside the given test context for each element of [ts].
+    * The test names will be generated from the stable properties of the elements. See [StableIdents].
+    */
+   suspend fun <T> withData(
+      ts: Iterable<T>,
+      test: suspend BehaviorSpecGivenContainerScope.(T) -> Unit
+   ) {
+      withData({ StableIdents.getStableIdentifier(it) }, ts, test)
+   }
+
+   /**
+    * Registers tests inside the given test context for each element of [ts].
+    * The test name will be generated from the given [nameFn] function.
+    */
+   suspend fun <T> withData(
+      nameFn: (T) -> String,
+      ts: Sequence<T>,
+      test: suspend BehaviorSpecGivenContainerScope.(T) -> Unit
+   ) = withData(nameFn, ts.toList(), test)
+
+   /**
+    * Registers tests inside the given test context for each element.
+    * The test name will be generated from the given [nameFn] function.
+    */
+   suspend fun <T> withData(
+      nameFn: (T) -> String,
+      first: T,
+      second: T,
+      vararg rest: T,
+      test: suspend BehaviorSpecGivenContainerScope.(T) -> Unit
+   ) = withData(nameFn, listOf(first, second) + rest, test)
+
+   /**
+    * Registers tests inside the given [FunSpecContainerScope] for each element of [ts].
+    * The test name will be generated from the given [nameFn] function.
+    */
+   suspend fun <T> withData(
+      nameFn: (T) -> String,
+      @BuilderInference ts: Iterable<T>,
+      @BuilderInference test: suspend BehaviorSpecGivenContainerScope.(T) -> Unit
+   ) {
+      ts.forEach { t -> addAnd(nameFn(t), false) { test(t) } }
+   }
```

Comment:
Why is `@BuilderInference` required here, but not on the [similar signature](https://github.com/kotest/kotest/pull/4599/files?diff=split&w=0#diff-0891129a199de5858e0d80f2f27ec463938bb8ab096088592560846b7756e15aR97) in `BehaviorSpecContextContainerScope`?

Reply:
Just a copy n paste thing from the existing withData.
I've removed them.